### PR TITLE
feat(detection): allow dotenv-style template files via suffix rule

### LINKED ISSUE #99: deny-read gate blocks dotenv-style template files (envrc sample/example, per-environment variants)
## Summary

The deny-read gate blocks template/sample variants of dotenv-style files that hold no secrets by design. The deny glob `.env*` matches every dotenv-style name, and the allowlist (`candidate_is_allowed()`) exempts only four exact basenames: `.env.example`, `.env.sample`, `.env.template`, `.env.dist`.

Common committed template names that are still blocked (false positives):

- `.envrc.sample`, `.envrc.example` (direnv templates)
- `.env.local.example`, `.env.production.sample`, `.env.test.example` (per-environment dotenv templates)

These are placeholder files meant to be committed and read; blocking them adds friction without protecting anything.

## Proposal

Extend the allowlist from four exact basenames to a suffix rule: allow any basename that starts with `.env` **and ends with** a template suffix — `.example`, `.sample`, `.template`, or `.dist`.

- Keeps the existing security property: the match stays on the basename shape, and a template *suffix* is required, so `.env`, `.envrc`, `.env.production` remain blocked.
- An agent renaming a real secrets file to earn the suffix (`cp .env .env.example`) is already possible with the current exact-name allowlist and is out of scope for this path gate; content-level scanning still applies elsewhere.
- The same allowlist is consulted by both the Read-path check and the Bash command token check (`bash_strip_allowed_tokens`), so both paths should honor the new rule consistently.

## Acceptance

- `.envrc.sample`, `.envrc.example`, `.env.local.example`, `.env.production.sample` pass the Read gate and the Bash token gate.
- `.env`, `.envrc`, `.env.production`, `.env.example.bak` stay blocked.
- Existing four exact names keep working.
- Tests cover both gates.

### PR DESCRIPTION
**Pull request by JeongJaeSoon**
## Summary

Extends the dotenv-template allowlist in `plugins/agent-guard/bin/agent-guard` (`candidate_is_allowed()`) from four exact basenames (`.env.example`, `.env.sample`, `.env.template`, `.env.dist`) to a suffix rule: any basename that starts with `.env` **and ends with** `.example`, `.sample`, `.template`, or `.dist` is exempt from the deny-read gate.

Closes #99

## What changed

- `candidate_is_allowed()` now strips the directory part first and matches the basename against anchored suffix globs (`.env*.example` etc.). Basename-only matching means the wildcard can never span a `/`, and the anchored suffix keeps the security property from the function's comment: only shapes that never hold secrets are exempted.
- Newly allowed (previously false positives): `.envrc.sample`, `.envrc.example`, `.env.local.example`, `.env.production.sample`, and similar committed templates.
- Still blocked: `.env`, `.envrc`, `.env.production` (no template suffix) and `.env.example.bak` (suffix not final).
- Both consumers of the allowlist are covered by the single edit: the Read-path gate (`candidate_matches_deny`) and the Bash token gate (`bash_strip_allowed_tokens`).
- Documented a pre-existing caveat on `bash_strip_allowed_tokens`: tokens are stripped textually without symlink resolution (the Read path re-checks the resolved target); the suffix rule widens the set of names that shape admits, so the comment now states it explicitly.
- 11 new test cases in `tests/run.sh` covering both gates: allow (`.envrc.sample`, `.env.local.example`, path-prefixed `config/.env.local.example`, `config/.envrc.sample`) and block (`.envrc`, `.env.example.bak`, `config/.env.local`) directions.

## Manual verification guide

Automated coverage: `make test` runs the full suite (354 tests, including the 11 new cases) with a mock gitleaks. What a human should spot-check by hand — live hook behavior with the real installed plugin:

1. Run the suite in the repo root:

   ```sh
   make test
   ```

   Expected: `passed: 354`, `failed: 0` (includes the new suffix-rule cases; grep the output for `envrc.sample`).

2. Exercise the hook binary directly — a newly allowed template (expect exit 0, no output):

   ```sh
   printf '%s' '{"tool_name":"Read","tool_input":{"file_path":".envrc.sample"}}' \
     | plugins/agent-guard/bin/agent-guard hook-pre-tool; echo "exit=$?"
   ```

3. A per-environment template through the Bash token gate (expect exit 0):

   ```sh
   printf '%s' '{"tool_name":"Bash","tool_input":{"command":"cat .env.local.example"}}' \
     | plugins/agent-guard/bin/agent-guard hook-pre-tool; echo "exit=$?"
   ```

4. Still-blocked names (expect exit 2 and a deny message on stderr for each):

   ```sh
   printf '%s' '{"tool_name":"Read","tool_input":{"file_path":".envrc"}}' \
     | plugins/agent-guard/bin/agent-guard hook-pre-tool; echo "exit=$?"
   printf '%s' '{"tool_name":"Read","tool_input":{"file_path":".env.example.bak"}}' \
     | plugins/agent-guard/bin/agent-guard hook-pre-tool; echo "exit=$?"
   printf '%s' '{"tool_name":"Bash","tool_input":{"command":"cat .env.example .env"}}' \
     | plugins/agent-guard/bin/agent-guard hook-pre-tool; echo "exit=$?"
   ```

5. In a live Claude Code session with the plugin installed (what automation can't cover): ask the agent to read a committed `.envrc.sample` in a real project — it should now succeed — and confirm reading a real `.env` is still denied.

## Notes

- No lint/shellcheck target exists in the Makefile or CI; the script's strict POSIX-sh style was preserved (case globs, no `local`).
- Renaming a real secrets file to earn the suffix (`cp .env .env.example`) was already possible with the exact-name allowlist and remains out of scope for this path gate; content-level scanning still applies elsewhere (see issue #99).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details>
<summary>code-flow usage JSON</summary>

```json
{
  "commits": [
    {
      "confidence": "best_effort",
      "model_totals": [
        {
          "cache_creation_ephemeral_1h_input_tokens": 209901,
          "cache_creation_input_tokens": 209901,
          "cache_read_input_tokens": 7191267,
          "input_tokens": 32722,
          "model": "claude-fable-5",
          "output_tokens": 67966,
          "total_context_tokens": 7501856,
          "turns": 52
        }
      ],
      "participants": [
        {
          "attribution": "exact",
          "cache_creation_ephemeral_1h_input_tokens": 209901,
          "cache_creation_input_tokens": 209901,
          "cache_read_input_tokens": 7191267,
          "input_tokens": 32722,
          "kind": "main",
          "model": "claude-fable-5",
          "name": "main",
          "output_tokens": 67966,
          "total_context_tokens": 7501856,
          "turns": 52
        }
      ],
      "recorded_at": "2026-07-06T18:19:52Z",
      "sha": "3c7b130d4f548619ac8fff9b04c0b2630122b73c",
      "subject": "feat(detection): allow dotenv-style template files via suffix rule",
      "totals": {
        "cache_creation_ephemeral_1h_input_tokens": 209901,
        "cache_creation_input_tokens": 209901,
        "cache_read_input_tokens": 7191267,
        "input_tokens": 32722,
        "output_tokens": 67966,
        "total_context_tokens": 7501856,
        "turns": 52
      }
    },
    {
      "confidence": "best_effort",
      "model_totals": [],
      "participants": [],
      "sha": "0c7daf0551573cb971f93f1121f196cfda7aa7b7",
      "status": "unmetered",
      "subject": "fix(detection): keep templates under deny-listed directories blocked",
      "totals": {
        "cache_creation_input_tokens": 0,
        "cache_read_input_tokens": 0,
        "input_tokens": 0,
        "output_tokens": 0,
        "total_context_tokens": 0
      }
    }
  ],
  "confidence": "best_effort",
  "metered_commits": 1,
  "pr": {
    "base": "main",
    "head": "feat/dotenv-template-suffix-allowlist",
    "number": 100
  },
  "schema": "code-flow-usage/1",
  "scope": "current_branch_pr",
  "source": "claude_code_hooks",
  "total_commits": 2,
  "totals": {
    "cache_creation_ephemeral_1h_input_tokens": 209901,
    "cache_creation_input_tokens": 209901,
    "cache_read_input_tokens": 7191267,
    "input_tokens": 32722,
    "output_tokens": 67966,
    "total_context_tokens": 7501856,
    "turns": 52
  },
  "totals_by_model": [
    {
      "cache_creation_ephemeral_1h_input_tokens": 209901,
      "cache_creation_input_tokens": 209901,
      "cache_read_input_tokens": 7191267,
      "input_tokens": 32722,
      "model": "claude-fable-5",
      "output_tokens": 67966,
      "total_context_tokens": 7501856,
      "turns": 52
    }
  ],
  "updated_at": "2026-07-06T18:45:52Z"
}
```

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of `.env*` file patterns so template-style filenames are recognized more consistently, including in nested paths.
  * Tightened blocking for sensitive filenames that don’t match the approved template-style suffixes.

* **Tests**
  * Expanded policy coverage for allowed template filenames, blocked secret names, and path-based edge cases.

### CODE REVIEW

File: plugins/agent-guard/bin/agent-guard [hunk 1]
```diff
@@ -836,14 +836,16 @@ resolve_path() {
...
+# Matched on the basename only, with the suffix anchored at the end, so a real
+# secret file (`.env`, `.envrc`, `.env.production`, `.env.example.bak`) is never
+# exempted by a crafted path. NOTE: a match here exempts the candidate from ALL
+# deny patterns, so the rule must only admit shapes that never hold secrets.
 candidate_is_allowed() {
-  case "$1" in
-    .env.example|.env.sample|.env.template|.env.dist) return 0 ;;
-    */.env.example|*/.env.sample|*/.env.template|*/.env.dist) return 0 ;;
+  case "${1##*/}" in
+    .env*.example|.env*.sample|.env*.template|.env*.dist) return 0 ;;
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard]: Thanks for catching this — fixed in commit 0c7daf0551573cb971f93f1121f196cfda7aa7b7.

`candidate_is_allowed()` now grants the template exemption only after confirming no parent path component matches a deny pattern: a new `candidate_parent_matches_deny()` helper walks the deny patterns against the candidate's parent path (`${1%/*}`) before the short-circuit fires. Since both the Read gate (`candidate_matches_deny`) and the Bash token gate (`bash_strip_allowed_tokens`) share `candidate_is_allowed()`, the fix covers both paths.

Your examples now stay blocked, verified by three new tests in `tests/run.sh`: a Read of a template inside a deny-listed directory, a Read with a deny-listed mid-path component, and a Bash `cat` of a template inside a deny-listed directory. Full suite: 357 passed / 0 failed.

Written by claude code assistant
